core/src/rules/conditional_directive_consolidator.rs

```diff
@@ -0,0 +1,401 @@
+    fn is_allowed_token(tokens: &[Token], token_index: usize) -> bool {
+        matches!(
+            tokens.get(token_index).map(Token::get_token_type),
+            Some(
+                TT::Identifier
+                    | TT::NumberLiteral(_)
+                    | TT::TextLiteral(TextLiteralKind::SingleLine)
+                    | TT::Comment(
+                        CommentKind::IndividualBlock
+                            | CommentKind::IndividualLine
+                            | CommentKind::InlineBlock
+                            | CommentKind::InlineLine
+                    )
+                    | TT::Op(OperatorKind::Dot),
+            )
+        )
+    }
+
+    fn expand_line(tokens: &[Token], line: &mut LogicalLine) -> Vec<usize> {
+        let line_tokens = line.get_tokens();
+        let (Some(&first_token), Some(&last_token)) = (line_tokens.first(), line_tokens.last())
+        else {
+            return vec![];
+        };
+
+        if last_token - first_token + 1 == line_tokens.len() {
+            // if this equality holds, the tokens are sequential and there is
+            // nothing to do
+            return vec![];
+        }
+
+        enum ConditionalState {
+            Outside,
+            AfterIf,
+            AfterElse,
+        }
+        use ConditionalState as CS;
+
+        let mut state = ConditionalState::Outside;
+        let mut directives: Vec<usize> = Vec::with_capacity(3);
+        let mut new_line_tokens = vec![first_token];
+
+        for (&prev, &current) in line.get_tokens().iter().tuple_windows() {
+            if current - prev > 1 {
+                let gap_start_tok = prev + 1;
+                let gap_end_tok = current - 1;
+                match (
+                    tokens.get(gap_start_tok).map(Token::get_token_type),
+                    tokens.get(gap_end_tok).map(Token::get_token_type),
+                ) {
+                    (
+                        Some(TT::ConditionalDirective(b_kind)),
+                        Some(TT::ConditionalDirective(e_kind)),
+                    ) => match &state {
+                        CS::Outside if b_kind.is_if() => {
+                            state = match e_kind {
+                                k if k.is_if() && gap_start_tok == gap_end_tok => CS::AfterIf,
+                                k if k.is_else() => CS::AfterElse,
+                                _ => return vec![],
+                            };
+                        }
+                        CS::AfterIf if b_kind.is_else() => {
+                            state = match e_kind {
+                                k if k.is_else() && gap_start_tok == gap_end_tok => CS::AfterElse,
+                                k if k.is_end() => CS::Outside,
+                                _ => return vec![],
+                            };
+                        }
+                        CS::AfterIf | CS::AfterElse if b_kind.is_end() && e_kind.is_end() => {
```

Comment:
This line needs the same treatment

```suggestion
                        CS::AfterIf | CS::AfterElse if b_kind.is_end() && gap_start_tok == gap_end_tok => {
```

without this, you get this formatting
```delphi
{$if}
A :=
    {$if} A {$else} B {$endif}
        {$endif};
```

instead of the intended
```delphi
{$if}
A :=
  {$if}
    A
  {$else}
    B
  {$endif}
{$endif}
    ;
```

Please add a test for this as well.